### ISSUE TITLE
The future of this project

### ISSUE THREAD

**Issue by evkaky** [JetBrains/kotlin-spark-api#220]
The activity of the project is super low (active PRs, issues etc). No roadmaps, no new features, almost nothing.
Is Jetbrains going to abandon this project at some point? If it is - why so?

**Comment by Jolanrensen** [comment]
Unfortunately the project at the moment has a low priority indeed. I took responsibility to keep it up to date, but my duties at Kotlin DataFrame need to come first. There is more user interest in that library and since Apache rejected the Kotlin API from becoming a first-class citizen, it's getting harder to justify my work here.
Also, keeping it up to date has become exceedingly hard, thanks to [Spark 3.4+](https://github.com/Kotlin/kotlin-spark-api/pull/218). I'm working on an a rebuild for supporting those Spark versions in my free time, so hopefully I can continue keeping the library up-to-date, however, new features are not planned.
That said, the project is open-source; anyone willing can contribute. It's what I did before joining JetBrains and I will be more than ready to help with contributions when needed.

**Comment by evkaky** [comment]
> Apache rejected the Kotlin API from becoming a first-class citizen

That's terrible. Deeply disappointed about this Apache's decision. Kotlin would fit for Spark jobs/notebooks much better with its DSLs than Python or Scala or SQL

**Comment by Jolanrensen** [comment]
@evkaky I agree! Unfortunately Spark is written in Scala with only Java in mind, so coop with Kotlin in the same repo is difficult (hence this API). That said, there's also unofficial support for [.NET and Spark](https://github.com/dotnet/spark) and that seems to be doing fine, so if we can keep this API working we should be good to go.

**Comment by evkaky** [comment]
@Jolanrensen BTW, event if Apache declined Kotlin as a first class API like Scala / Java / Python, can we (JetBrains ?) just utilise  this all-new Spark's "connect" feature to built a solid API?
https://spark.apache.org/docs/latest/spark-connect-overview.html

**Comment by Jolanrensen** [comment]
@evkaky It's actually something I'm looking into for [3.4+ support](https://github.com/Kotlin/kotlin-spark-api/pull/218). One downside atm is that I cannot get spark-connect itself to work reliably at all. Running a Spark connect server with spark 3.5.1 and Scala 2.13 just breaks completely and when I target 2.12 I cannot get their simple standalone application sample to work due to various NoSuchMethodErrors etc. It would be very helpful if someone could give me a working sample with Spark-connect to check whether my 3.4+ implementation would actually work with it.

**Comment by andrija-s** [comment]
Umm, please live.

**Comment by Jolanrensen** [comment]
@andrija-s Just begging won't bring updates to the project. A lot of time and hard work does. I'm working full-time on [Kotlin DataFrame](https://github.com/Kotlin/dataframe). At the moment, we just don't have the time or resources needed to continue the 3.4+ path I started [here](https://github.com/Kotlin/kotlin-spark-api/issues/195). Plus, undoubtedly, Spark will break more stuff in the future, which will require even more time. Time which we may have in the future, but just not right now.

I don't want it to die either. When I found this project some time ago and I needed it myself, I contributed a lot of work to improve it. But so could you :) or anyone from the community. If you think the project is valuable enough and you're willing to invest the time and work I originally did, I'm not stopping you, in fact, I'm encouraging it :)

**Comment by PeterAttardo** [comment]
They recently [announced Spark 4.0](https://www.databricks.com/blog/introducing-apache-spark-40), which according to the release notes includes big jumps in Spark Connect -- what they describe as "high feature parity with Spark Classic." To go along with this they've developed Spark Connect libraries for Go, Rust, and Swift. If Spark Connect is as clean and feature rich in 4.0 as they're advertising, it may be worth skipping 3.4 compatibility entirely and revisiting the Spark Connect path to a Kotlin API compatible with 4.0.

**Comment by Jolanrensen** [comment]
@PeterAttardo I agree that might be an easier thing to do, especially if we're able to wrap the Scala Spark Connect API. From the Kotlin side, we'd only have to make the API a bit more Kotlin friendly, so support for data classes, auto-encoders, some extension functions, DSL helpers, and Tuple helpers. I also noticed using Spark Connect in notebooks was easier to get working than Spark Classic because Spark tries to be less clever with serialization of objects on-the-fly.

That said, it's way harder for a user to get started with Spark Connect. Unlike Spark Classic, it requires a server to run, which makes testing and building this project harder, as well as providing easy examples.

**Comment by Jolanrensen** [comment]
While to bump this library to the next Spark versions,
this proved to be near impossible using the current library design. In short: a whole rewrite is needed, for which the current maintainers of this project simply don't have the time.
That's why I'm very happy to say Yerold Sanabria Rios and Eveliina Vesala, for their [Bachelor's thesis](https://www.theseus.fi/handle/10024/926324), decided to research
what would be the best way to create a Kotlin Spark API in the Spark Connect age. They welcome forks of their [GitHub repo](https://github.com/eveliinavesala/spark-connect-kotlin/tree/main),
so have a look if you're interested! And maybe together we can continue using Spark in the language we love most <3.
